Sources/AblyAssetTrackingCore/ResultHandler.swift

```diff
@@ -1,10 +1,68 @@
-/**
- Completion handler for operations
- */
-public typealias ResultHandler<T: Any> = (Result<T, ErrorInformation>) -> Void
+import Foundation
 
-public extension Result where Success == Void {
-    static var success: Result {
+public typealias ResultHandler = (ATResult) -> Void
+
+@objc
+public protocol Resultable {
+    var success: Any? { get }
+    var failure: ErrorInformation? { get }
+}
+
+@objcMembers
+public class ATResult: NSObject, Resultable {
+    public let success: Any?
+    public let failure: ErrorInformation?
+    
+    public static var success: ATResult {
+        return .init(success: Void(), failure: nil)
+    }
+    
+    public static func success(_ value: Any?) -> ATResult {
+        return .init(success: value, failure: nil)
+    }
+    
+    public static func failure(_ value: ErrorInformation?) -> ATResult {
+        return .init(success: nil, failure: value)
+    }
+    
+    public init(success: Any? = nil, failure: ErrorInformation? = nil) {
+        self.success = success
+        self.failure = failure
+    }
+    
+    public override init() {
+        self.success = nil
+        self.failure = nil
+        
+        super.init()
+    }
+}
+
+public extension ATResult {
+    enum Unwrapped<T> {
+        case success(T)
+        case failure(ErrorInformation)
+    }
+    
+    var enumUnwrap: Unwrapped<Void> {
+        if let failure = failure {
+            return .failure(failure)
+        }
+        
         return .success(Void())
     }
+    
+    func enumUnwrap<T>(_:T.Type) -> Unwrapped<T> {
```

Comment:
I've meant method name `unwrapped()` here

Reply:
I change to `unwrap`, IMO imperative is better
2174f9efce35991624694e40798cb7e24fc44acd

Reply:
I changed to imperative `unwrap(...)` here 2174f9efce35991624694e40798cb7e24fc44acd and I used Swift Result instead of custom enum, as @lawrence-forooghian recommended [here](https://github.com/ably/ably-asset-tracking-swift/pull/216#discussion_r761454307)

Reply:
`unwrap()` is a function name, `unwrapped` - property name (fit for function too). Also can't it be just one function with `T = Void` by default?

Reply:
it's not possible to set the default generic type, of course, I could create convenience method like `unwrap()` and in the body calls `unwrap(Void())`

but for now, I'll not change it because the idea with `Result` will probably change